ruby/ql/lib/codeql/ruby/ast/Pattern.qll

```diff
@@ -97,3 +97,322 @@ class TuplePattern extends Pattern, TTuplePattern {
+  ConstantReadAccess getClass() { toGenerated(result) = g.getClass() }
+
+  /**
+   * Gets the `n`th element of this list pattern's prefix, i.e. the elements `1, ^two, 3`
+   * in the following examples:
+   * ```
+   * in [ 1, ^two, 3 ]
+   * in [ 1, ^two, 3, ]
+   * in [ 1, ^two, 3, *, 4 , 5]
+   * in [ 1, ^two, 3, *more]
+   * ```
+   */
+  CasePattern getPrefixElement(int n) {
+    toGenerated(result) = g.getChild(n) and
+    (
+      n < this.restIndex()
+      or
+      not exists(restIndex())
+    )
+  }
+
+  /**
+   * Gets the `n`th element of this list pattern's suffix, i.e. the elements `4, 5`
+   * in the following examples:
+   * ```
+   * in [ *, 4, 5 ]
+   * in [ 1, 2, 3, *middle, 4 , 5]
+   * ```
+   */
+  CasePattern getSuffixElement(int n) { toGenerated(result) = g.getChild(n + this.restIndex() + 1) }
+
+  /**
+   * Gets the variable of the rest token, if any. For example `middle` in `the following array pattern.
+   * ```rb
+   * [ 1, 2, 3, *middle, 4 , 5]
+   * ```
+   */
+  LocalVariableWriteAccess getRestVariable() {
+    toGenerated(result) = g.getChild(restIndex()).(Ruby::SplatParameter).getName()
+  }
+
+  /**
+   * Holds if this pattern permits any unmatched remaining elements, i.e. the pattern does not have a trailing `,`
+   * and does not contain a rest token (`*` or `*name`) either.
+   */
+  predicate allowUnmatchedElements() { not exists(this.restIndex()) }
+
+  private int restIndex() { g.getChild(result) instanceof Ruby::SplatParameter }
+
+  final override string getAPrimaryQlClass() { result = "ArrayPattern" }
+
+  final override string toString() { result = "[ ..., * ]" }
+
+  final override AstNode getAChild(string pred) {
+    result = super.getAChild(pred)
+    or
+    pred = "getClass" and result = this.getClass()
+    or
+    pred = "getPrefixElement" and result = this.getPrefixElement(_)
+    or
+    pred = "getSuffixElement" and result = this.getSuffixElement(_)
+    or
+    pred = "getRestVariable" and result = this.getRestVariable()
+  }
+}
+
+/**
+ * A find pattern, for example:
+ * ```rb
+ * in [*, "a", Integer => x, *]
+ * in List[*init, "a", Integer => x, *tail]
+ * in List[*, "a", Integer => x, *]
+ * ```
+ */
+class FindPattern extends CasePattern, TFindPattern {
+  private Ruby::FindPattern g;
+
+  FindPattern() { this = TFindPattern(g) }
+
+  /** Gets the class this pattern matches objects against, if any. */
+  ConstantReadAccess getClass() { toGenerated(result) = g.getClass() }
+
+  /** Gets the `n`th element of this list pattern. */
+  CasePattern getElement(int n) { toGenerated(result) = g.getChild(n + 1) }
+
+  /** Gets an element of this list pattern. */
+  CasePattern getAnElement() { result = this.getElement(_) }
+
+  /**
+   * Gets the variable for the prefix of this list pattern, if any. For example `init` in:
+   * ```rb
+   * in List[*init, "a", Integer => x, *tail]
+   * ```
+   */
+  LocalVariableWriteAccess getPrefix() {
```

Comment:
Again, I would like `LocalVariable` return type, and perhaps rename to `getPrefixVariable`.

Reply:
Renamed the method to getPrefixVariableAccess.